File: library/src/main/java/com/yy/codex/uikit/UISlider.java
package com.yy.codex.uikit;

import android.content.Context;
import android.graphics.Outline;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.RequiresApi;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewOutlineProvider;

import com.yy.codex.foundation.NSLog;

/**
 * Created by adi on 17/1/19.
 */

public class UISlider extends UIControl {

    private UIView thumbView;
    private UIView trackView;
    private UIView trackPassedView;
    private double value; // range: 0.0 ~ 1.0
    private UISliderCallback callback;

    private double thumbRadius = 30.0;

    private boolean active;


    public UISlider(@NonNull Context context, @NonNull View view) {
        super(context, view);
    }

    public UISlider(@NonNull Context context) {
        super(context);
    }

    public UISlider(@NonNull Context context, @NonNull AttributeSet attrs) {
        super(context, attrs);
    }

    public UISlider(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public UISlider(@NonNull Context context, @NonNull AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    private void defaultValue(){
        value = 0.5;
        thumbRadius = 30.0;
    }

    @Override
    protected void init() {
        super.init();
        defaultValue();

        trackView = new UIView(getContext());
        trackView.setWantsLayer(true);
        trackView.getLayer().setBackgroundColor(new UIColor(0xb7/255.0, 0xb7/255.0, 0xb7/255.0, 1));
        trackView.getLayer().setCornerRadius(2);

        trackPassedView = new UIView(getContext());
        trackPassedView.setWantsLayer(true);
        trackPassedView.getLayer().setBackgroundColor(new UIColor(0x10/255.0, 0x6a/255.0, 1, 1));
        trackPassedView.getLayer().setCornerRadius(2);

        thumbView = new UIView(getContext());
        thumbView.setWantsLayer(true);
        thumbView.getLayer().setShadowX(2).setShadowY(2).setShadowRadius(0.5).setShadowColor(new UIColor(.3, .3, .3, .2));
        thumbView.getLayer().setCornerRadius(thumbRadius/2.0).setBorderWidth(0.5).setBorderColor(new UIColor(0x00/255.0, 0x00/255.0, 0x00/255.0, 0.15));
        thumbView.getLayer().setBackgroundColor(UIColor.whiteColor);

        addSubview(trackView);
        addSubview(trackPassedView);
        addSubview(thumbView);
    }


    @Override
    public void layoutSubviews() {
        super.layoutSubviews();
        NSLog.warn(getFrame());

        double frameW = getFrame().getSize().getWidth();
        trackView.setFrame(new CGRect(0, 14, frameW, 4));
        trackPassedView.setFrame(new CGRect(0, 14, frameW * value, 4));
        thumbView.setFrame(new CGRect(frameW * value - thumbRadius/2.0, 2, thumbRadius, thumbRadius));
    }

    @Override
    protected void onEvent(Event event) {
        super.onEvent(event);
        switch (event){
            case TouchUpOutside:
            case TouchUpInside:
                // reset active = false
                 break;
            case TouchDown:
                // is pointInSide?
                // if YES
                // ... active = true
                // ...
//            case Touch
        }
    }

    

    public void onSlide(UISliderCallback callback){
        this.callback = callback; // callback.handle(this.value);
    }

}

File: library/src/main/java/com/yy/codex/uikit/UISlider.kt
package com.yy.codex.uikit

import android.content.Context
import android.os.Build
import android.support.annotation.RequiresApi
import android.util.AttributeSet
import android.view.View
import com.yy.codex.foundation.lets

/**
 * Created by adi on 17/1/19.
 */

class UISlider : UIControl {

    /* initialize */

    constructor(context: Context, view: View) : super(context, view) {}
    constructor(context: Context) : super(context) {}
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {}
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
    }

    override fun init() {
        super.init()
        initTrackView()
        addSubview(trackView)
        initProgressView()
        addSubview(progressView)
        initThumbView()
        addSubview(thumbView)
        resetViewFrames()
    }

    override fun tintColorDidChanged() {
        super.tintColorDidChanged()
        lets(progressView, tintColor, { progressView, tintColor ->
            progressView.layer.backgroundColor = tintColor
        })
    }

    override fun intrinsicContentSize(): CGSize {
        return CGSize(0.0, 44.0)
    }

    /* appearance */

    private lateinit var trackView: UIView

    private fun initTrackView() {
        trackView = UIView(context)
        trackView.wantsLayer = true
        trackView.layer.backgroundColor = UIColor(0xb7 / 255.0, 0xb7 / 255.0, 0xb7 / 255.0, 1.0)
        trackView.layer.cornerRadius = 1.0
    }

    private lateinit var thumbView: UIView

    private fun initThumbView() {
        thumbView = UIView(context)
        thumbView.wantsLayer = true
        thumbView.layer.shadowX = 2.0
        thumbView.layer.shadowY = 2.0
        thumbView.layer.shadowRadius = 2.0
        thumbView.layer.shadowColor = UIColor(.3, .3, .3, .2)
        thumbView.layer.cornerRadius = (kThumbRadius / 2.0)
        thumbView.layer.borderWidth = 0.5
        thumbView.layer.borderColor = UIColor(0x00 / 255.0, 0x00 / 255.0, 0x00 / 255.0, 0.15)
        thumbView.layer.backgroundColor = UIColor.whiteColor
        thumbView.layer.wantsEnlargerLayer()
    }

    private lateinit var progressView: UIView

    private fun initProgressView() {
        progressView = UIView(context)
        progressView.wantsLayer = true
        progressView.layer.backgroundColor = if (tintColor != null) tintColor!! else UIColor(0x10 / 255.0, 0x6a / 255.0, 1.0, 1.0)
        progressView.layer.cornerRadius = 1.0
    }

    var progressValue = 0.5
        set(value) {
            if (value < 0.0){
                field = 0.0
            }
            else if (value > 1.0){
                field = 1.0
            }
            else {
                field = value
            }
            resetViewFrames()
        }

    override fun layoutSubviews() {
        super.layoutSubviews()
        resetViewFrames()
    }

    private fun resetViewFrames() {
        val frameW = frame.size.width
        trackView.frame = CGRect(0.0, 15.0, frameW - 6, 2.0)
        progressView.frame = CGRect(0.0, 15.0, (frameW - kThumbRadius - 4) * progressValue, 2.0)
        thumbView.frame = CGRect((frameW - kThumbRadius - 4) * progressValue, 1.0, kThumbRadius, kThumbRadius)
    }

    private var sliderTracking = false

    override fun onLongPressed(sender: UILongPressGestureRecognizer) {
        super.onLongPressed(sender)
        if (sender.state == UIGestureRecognizerState.Began) {
            if (UIViewHelpers.pointInside(thumbView, sender.location(thumbView))) {
                sliderTracking = true
            }
        }
        else if (sender.state == UIGestureRecognizerState.Changed && sliderTracking) {
            this.progressValue = Math.max(0.0, Math.min(1.0, sender.location(this).x / frame.width))
            onEvent(Event.ValueChanged)
        }
        else if (sender.state == UIGestureRecognizerState.Ended) {
            sliderTracking = false
        }
    }

    companion object {

        private val kThumbRadius = 30.0

    }

}
